feat: Added the PipelineJob.from_pipeline_func method

Ark-kun opened the pull request:
The new factory method reduces the pipeline submission boilerplate to absolute minimum.
```python
aiplatform.PipelineJob.from_pipeline_func(training_pipeline).submit()
```

What it does:

1. Compiles pipeline
2. Provides sensible default values for the pipeline display name, job_id, context etc.
3. Generates GCS directory for the pipeline output artifacts if needed
4. Creates the GCS bucket for the artifacts if it does not exist. (And gives the Pipelines service account the required permissions)

Example usage:

```python
def training_pipeline(number_of_epochs: int = 10):
    train_op(
        number_of_epochs=number_of_epochs,
        learning_rate="0.1",
    )

job = aiplatform.PipelineJob.from_pipeline_func(training_pipeline)
job.submit()
```
---
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-aiplatform/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

sasha-gitg commented on google/cloud/aiplatform/pipeline_jobs.py:
```diff
@@ -756,3 +758,147 @@ def clone(
+        """
+
+        # Importing the KFP module here to prevent import errors when the kfp package is not installed.
+        from kfp.v2 import compiler as compiler_v2
```
try/except and raise a more informative exception

If the user ended up in a scenario where kfp wasn't installed, I assume they would be very confused.

Ark-kun replied:
Done.

sasha-gitg commented on google/cloud/aiplatform/utils/gcs_utils.py:
```diff
@@ -25,7 +25,7 @@
 from google.cloud import storage
```
bump copyright to 2022

Ark-kun replied:
Done.

sasha-gitg commented on google/cloud/aiplatform/utils/gcs_utils.py:
```diff
@@ -163,3 +163,52 @@ def stage_local_data_in_gcs(
+    project: Optional[str] = None,
+    location: Optional[str] = None,
+    credentials: Optional[auth_credentials.Credentials] = None,
+):
```
docstring

Ark-kun replied:
Done.

sasha-gitg commented on google/cloud/aiplatform/utils/gcs_utils.py:
```diff
@@ -163,3 +163,52 @@ def stage_local_data_in_gcs(
     return staged_data_uri
+
+
+def create_gcs_directory_for_pipeline_artifacts(
```
this is great

Ark-kun replied:
Thank you!

P.S. I've decided to split this into a two-stage approach. It's better to not perform any changes in constructor or factory method. Now the bucket creation is delayed until the user calls `job.submit()`.

Ark-kun commented:
Hello.

I think I've implemented and resolved all feedback. I've also added the new method to the integration tests and unit tests. (I'm not sure the integration tests are checked during the presubmit).

sasha-gitg commented on tests/system/aiplatform/test_pipeline_job.py:
```diff
@@ -0,0 +1,61 @@
+
+
+@pytest.mark.usefixtures("tear_down_resources")
+class TestExperiments(e2e_base.TestEndToEnd):
```
```suggestion
class TestPipelineJob(e2e_base.TestEndToEnd):
```

Ark-kun replied:
Done.

sasha-gitg commented on tests/system/aiplatform/test_experiments.py:
```diff
@@ -276,13 +275,9 @@ def trainer(
-        )
-
-        job = aiplatform.PipelineJob(
+        job = aiplatform.PipelineJob.from_pipeline_func(
```
Preference to leave this test as-is since it's testing a different `PipelineJob` path than the test_pipeline_job.py test added in this PR.

Ark-kun replied:
Reverted

Ark-kun commented on google/cloud/aiplatform/utils/gcs_utils.py:
```diff
@@ -163,3 +163,103 @@ def stage_local_data_in_gcs(
+            project=project,
+            location=location,
+        )
+        # Giving the service account read and write access to teh new bucket
```
Thanks.

Ark-kun commented on google/cloud/aiplatform/pipeline_jobs.py:
```diff
@@ -776,3 +792,121 @@ def clone(
+
+        automatic_display_name = (
+            pipeline_func.__name__.replace("_", " ")
+            + " "
```
Done.
